### PR TITLE
decorate Selection.apply to ignore empty atomgroups

### LINKED ISSUE #2765: Cannot select atoms with StringSelection from empty AtomGroup
## Expected behavior ##

`empty_atomgroup.select_atoms('name anything')` returns an empty AtomGroup.

## Actual behavior ##

A TypeError is raised.

## Code to reproduce the behavior ##

```python
In [1]: import MDAnalysis as mda

In [2]: from MDAnalysis.tests.datafiles import PSF

In [3]: mda.__version__
Out[3]: '2.0.0-dev0'

In [4]: u = mda.Universe(PSF)

In [5]: ag = u.select_atoms('resname blahblah')

In [6]: len(ag)
Out[6]: 0

In [7]: ag.select_atoms('name O')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-c43a12bdaf29> in <module>
----> 1 ag.select_atoms('name O')

~/pydev/mdanalysis/package/MDAnalysis/core/groups.py in select_atoms(self, sel, *othersel, **selgroups)
   2861             # Apply the first selection and sum to it
   2862             atomgrp = sum([sel.apply(self) for sel in selections[1:]],
-> 2863                           selections[0].apply(self))
   2864         return atomgrp
   2865

~/pydev/mdanalysis/package/MDAnalysis/core/selection.py in apply(self, group)
    517         for val in self.values:
    518             values = getattr(group, self.field)
--> 519             mask |= [fnmatch.fnmatch(x, val) for x in values]
    520         return group[mask].unique
    521

TypeError: ufunc 'bitwise_or' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`) 2.0.0-dev0
- Which version of Python (`python -V`)? 3.7
- Which operating system? MacOS

### PR DESCRIPTION
**Pull request by lilyminium**
Fixes #2765 

Changes made in this Pull Request:
 - if an empty AtomGroup is passed to Selection.apply, just return it

I decorated `Selection.apply` in `_Selectionmeta.__init__`, which I baselessly feel is a bit iffy. Any opinions on if I should move it somewhere else?

Edit: alternately it could be expensive to have so many ifs, I could just do the `StringSelection` and sphzone/layer ones individually

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/core/selection.py [hunk 1]
```diff
@@ -156,6 +156,16 @@ def apply(self, group):
 
         return group.universe.atoms[idx]
 
+def return_empty_on_apply(func):
```
Comment (richardjgowers) [inline package/MDAnalysis/core/selection.py:159]: Iirc this kills docstrings so there's a `functools.decorate` or something to preserve docstrings after wrapping

**Comment by richardjgowers** [comment]
I like the fix but it's a little hacky to put the decorator application in the `_SelectionMeta` - really the Selection going through shouldn't be modified, so maybe it's better to explicitly include the `@emptyAGguard` decorator onto each `apply` method

**Comment by lilyminium** [comment]
I added the decorator to the apply methods the geometric and string selection classes instead.

**Comment by orbeckst** [comment]
This PR did not include a CHANGELOG comment. 

Further discussion in https://github.com/MDAnalysis/mdanalysis/issues/2765#issuecomment-654519327

**Comment by lilyminium** [comment]
Yes, sorry, we forgot; the CHANGELOG entry was updated in the next merge to develop (by @IAlibay iirc)

**Pull request merged**